services/code_hosting/git_hub_service.rb

```diff
@@ -1,131 +1,409 @@
+      #          "Description of what the lane does"
+      #        ],
+      #        "actions": [
+      #          {
+      #            "action": "gym",
+      #            "parameters": {
+      #              "skip_package_ipa": false,
+      #              "clean": true,
+      #              "project": "../fastlane-ci-demoapp.xcodeproj",
+      #              "scheme": "fastlane-ci-demoapp"
+      #            }
+      #          }
+      #        ],
+      #        "private": false
+      #      }
+      #    }
+      #  }
+      #
+      # @param repo_url [String]
+      # @param branch [String]
+      # @param path [String]
+      # @param provider_credential [GithubProviderCredential]
+      # @return [Hash<String, Hash>] being the key the name of the platform (:ios, :android, :no_platform) and the Hash the underlying lane (name, actions).
+      def peek_fastfile_configuration(repo_url: nil, branch: nil, provider_credential: nil, path: self.temp_path, cache: true)
+        repo = repo_from_url(repo_url)
+        return self.cache[[repo, branch].join("/")] if cache && self.cache && !self.cache[[repo, branch].join("/")].nil? && self.cache[[repo, branch].join("/")].kind_of?(Hash)
+        path = File.join(path, repo, branch)
+        begin
+          git_path = File.join(path, repo.split("/").last)
+          # This triggers the check of an existing repo in the given path,
+          # we recover from the error making the clone and checkout
+          Git.open(git_path)
+          fastfile_path = self.fastfile_path(root_path: git_path)
+          fastfile = Fastlane::FastfileParser.new(path: fastfile_path)
+          fastfile_config = {}
+          # TODO: FastfileParser provides a nice and clean tree of the Fastfile given,
+          # but there's a specific case when there might be lanes or actions out of
+          # a platform scope, in those cases the key is nil. We sanitize that so the user
+          # still can select lanes without platform. But this should be done from the
+          # FastfileParser side.
+          fastfile.tree.each_key do |key|
+            if key.nil?
+              fastfile_config[:no_platform] = fastfile.tree[key]
+            else
+              fastfile_config[key.to_sym] = fastfile.tree[key]
+            end
+          end
+          fastfile_json_path = File.join(path, "fastfile.json")
+          FileUtils.touch(fastfile_json_path) unless File.exist?(fastfile_json_path) && !File.zero?(fastfile_json_path)
+          File.write(fastfile_json_path, JSON.pretty_generate(fastfile_config))
+          self.cache[[repo, branch].join("/")] = fastfile, fastfile_config
+          return fastfile, fastfile_config
+        rescue ArgumentError
+          self.clone(
+            repo_url: repo_url,
+            branch: branch,
+            provider_credential: provider_credential,
+            path: path
+          )
+          fastfile, fastfile_config = self.peek_fastfile_configuration(
```

Comment:
this feels like we're just asking for infinite recursion. Maybe we should add an escape hatch somehow.

Reply:
So the only event that triggers the `ArgumentError` is the `Git.open` method, when the git does not exists. So yeah, this method will try to clone the repo until is really being opened without issues, maybe I can just add a `retry_count` that raises whenever after `n_retries` the clone is not being made.

Reply:
Oooh I like that